### DISCUSSION TITLE
Reading multiple trees using uproot.iterate()

### DISCUSSION THREAD

**Discussion by ligerlac** [scikit-hep/uproot5#1299] [Q&A] [answered]
Assume a dataset spread across many files and you want to read `branch_a_1` and `branch_a_2` from `tree_a` and `branch_b` from `tree_b`. The baseline approach would be:
```
for fn in glob.glob('files*.root'):
    with uproot.open(fn) as f:
        a_1 = f['tree_a/branch_a_1']
        a_2 = f['tree_a/branch_a_2']
        b = f['tree_b/branch_b']
        do_something ...
```
How can this be achieved with uproot.iterate()? I came up with this non-optimal work-around:
```
for tup in zip(
    uproot.iterate('files*.root:tree_a', ['branch_a_1', 'branch_a_2']),
    uproot.iterate('files*.root:tree_b', ['branch_b'])
):
    arr = ak.zip({f: a[f] for a in tup for f in ak.fields(a)})
    a_1 = arr['branch_a_1']
    a_2 = arr['branch_a_1']
    b = arr['branch_b']
    do_something ...
```
Is there a more elegant solution?

**Comment by jpivarski** [comment]
What you've described is a reasonable way to go about it. You probably want to set `step_size` in [uproot.iterate](https://uproot.readthedocs.io/en/latest/uproot.behaviors.TBranch.iterate.html) to an integer, to ensure that the arrays can always be [ak.zip](https://awkward-array.org/doc/main/reference/generated/ak.zip.html)ped together (ensure that they broadcast, by having the same number of entries).

**Reply by ligerlac** [reply]
Thanks for the swift answer, @jpivarski. The thing with step_size is a good point. I assume the work-around will result in opening each file twice? At least that would explain the ~20% longer run time I observed compared to the baseline approach.

**Reply by jpivarski** [reply, marked as answer]
I missed the fact that it was the same files, but different TTrees. If you want to ensure that you open the files only once, use [uproot.TTree.iterate](https://uproot.readthedocs.io/en/latest/uproot.behaviors.TTree.TTree.html#iterate) instead of [uproot.iterare](https://uproot.readthedocs.io/en/latest/uproot.behaviors.TBranch.iterate.html). This involves more bookkeeping, as you have to [uproot.open](https://uproot.readthedocs.io/en/latest/uproot.reading.open.html) each file manually.
